iree/compiler/Dialect/Stream/Transforms/test/elide_async_copies.mlir

```diff
@@ -0,0 +1,122 @@
+// RUN: iree-opt -split-input-file -iree-stream-elide-async-copies %s | IreeFileCheck %s
+
+// Tests that a normal clone-on-multiple-uses pattern has the last clone elided.
+// This is what the -iree-stream-materialize-copy-on-write pass generates and
+// expects us to clean up.
+
+// CHECK-LABEL: @multiUseTiedOperand
+func @multiUseTiedOperand(%size: index) -> (!stream.resource<*>, !stream.resource<*>) {
+  %c0 = arith.constant 0 : index
+  %c128 = arith.constant 128 : index
+  %c256 = arith.constant 256 : index
+  %c123_i32 = arith.constant 123 : i32
+  %c456_i32 = arith.constant 456 : i32
+  %c789_i32 = arith.constant 789 : i32
+  // CHECK: %[[SPLAT:.+]] = stream.async.splat
+  %splat = stream.async.splat %c123_i32 : i32 -> !stream.resource<*>{%size}
+  // CHECK: %[[CLONE0:.+]] = stream.async.clone %[[SPLAT]]
+  %clone0 = stream.async.clone %splat : !stream.resource<*>{%size} -> !stream.resource<*>{%size}
+  // CHECK: %[[FILL0:.+]] = stream.async.fill %c456_i32, %[[CLONE0]]
+  %fill0 = stream.async.fill %c456_i32, %clone0[%c0 to %c128 for %c128] : i32 -> %1 as !stream.resource<*>{%size}
+  // CHECK-NOT: stream.async.clone
+  %clone1 = stream.async.clone %splat : !stream.resource<*>{%size} -> !stream.resource<*>{%size}
+  // CHECK: %[[FILL1:.+]] = stream.async.fill %c789_i32, %[[SPLAT]]
+  %fill1 = stream.async.fill %c789_i32, %clone1[%c128 to %c256 for %c128] : i32 -> %3 as !stream.resource<*>{%size}
+  // CHECK: return %[[FILL0]], %[[FILL1]]
+  return %fill0, %fill1 : !stream.resource<*>, !stream.resource<*>
+}
+
+// -----
+
+// Tests a copy of a by-value function argument gets elided.
+// Since the caller passes in the last live reference the callee is allowed to
+// mutate the memory in-place.
```

Comment:
Thanks for all these clear test cases and comments, they make it pretty easy to follow the intent!